llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp

```diff
@@ -4138,9 +4138,10 @@ void SelectionDAGBuilder::visitAlloca(const AllocaInst &I) {
                                           APInt(IntPtr.getScalarSizeInBits(),
                                                 TySize.getKnownMinValue())));
   else
-    AllocSize =
-        DAG.getNode(ISD::MUL, dl, IntPtr, AllocSize,
-                    DAG.getConstant(TySize.getFixedValue(), dl, IntPtr));
+    AllocSize = DAG.getNode(ISD::MUL, dl, IntPtr, AllocSize,
+                            DAG.getConstant(APInt(IntPtr.getScalarSizeInBits(),
```

Comment:
Rather than constructing your own APInt, I think it would be clearer to use DAG.getZExtOrTrunc to the IntPtr type

Reply:
It seems to me that `getZExtOrTrunc` takes an `SDValue` as its input. `TySize` here is not a `Value *`, but a constant, so I am not sure how to create an `SDValue` out of it, other than using `DAG.getConstant` (which is crashing because `getConstant` does not like oversized constants). That is why an `APInt` is used, similar to the `isScalable` case above. A few lines above (https://github.com/llvm/llvm-project/pull/71472/files#diff-34ee9626bb04949852b8ca80ba74ba7e9a43d7875410dcf993d36720f3ea6058R4132), the code performs the check we want on `AllocaSize`. I could add an additional check there for `TySize` as well, and I was limited by the fact that `TySize` is not a `Value *`. 

I must be missing some APIs to do such a thing. Could you kindly point out what I am missing here? Thanks so much!

Reply:
The code immediately above already does this for the scalable vector case.

```
 EVT IntPtr = TLI.getPointerTy(DAG.getDataLayout(), I.getAddressSpace());
  if (AllocSize.getValueType() != IntPtr)
    AllocSize = DAG.getZExtOrTrunc(AllocSize, dl, IntPtr);
```

You just need to produce the IntVT based on the size (e.g. MVT::getIntegerVT(IntPtr.getScalarSizeInBits()). You also don't need the type check here, getZExtOrTrunc does that for you

Reply:
Thanks so much @arsenm ! The code is revised to take advantage of `getZExtOrTrunc` and to avoid creating an extra `APInt`. I still had some issues trying to figure out a good way to compute `TySize`'s bit width, but since it is represented by a `uint64_t`, the bitwidth is hardcoded to 64. Using `IntPtr.getScalarSizeInBits` does not work for me because it is too narrow for the big constants under our concern.